File: individual/client/src/main/java/requests/AckPacket.java
package requests;

import fields.OpCode;

import java.nio.ByteBuffer;

public class AckPacket extends Packet {

    private final short blockNumber;

    public AckPacket(short blockNumber) {
        super(OpCode.ACK);
        this.blockNumber = blockNumber;
    }

    @Override
    public int size() {
        return Character.BYTES + Character.BYTES;
    }

    @Override
    public byte[] serialize() {
        ByteBuffer buf = ByteBuffer.allocate(size());
        buf.putShort(getOpCode());
        buf.putChar((char)blockNumber);
        return buf.array();
    }
}

File: udp/src/main/kotlin/ru/hse/anstkras/tftp/packet/AckPacket.kt
package ru.hse.anstkras.tftp.packet

import java.nio.ByteBuffer

class AckPacket(private val blockNum: Int) : Packet {

    //    2 bytes     2 bytes
    //    ---------------------
    //    | Opcode |   Block #  |
    //    ---------------------
    @ExperimentalUnsignedTypes
    override fun getBytesRepresentation(): ByteBuffer {
        val buffer = ByteBuffer.allocate(4)
        buffer.putShort(4)
        buffer.putShort(blockNum.toUShort().toShort())
        return buffer
    }

    companion object : Parsable<Packet> {
        override fun parse(byteBuffer: ByteBuffer): Packet {
            val blockNum = byteBuffer.short
            return AckPacket(blockNum.toInt())
        }
    }
}